syntax = "proto3";

service managementServer {
    // 数据服务器上线后向管理服务器登记注册
    rpc serverOnline(serverInfo) returns (ma_reply) {}
    // 数据服务器下线后向管理服务器登记注销
    rpc serverOffline(serverId) returns (ma_reply) {}
    // 获取当前在线数据服务器信息
    rpc getServerList(empty) returns (serverList) {}
    // 给文件上锁
    rpc lockFile(lockInfo) returns (lockReply) {}
    // 给文件解锁
    rpc unlockFile (lockInfo) returns (lockReply) {}
    // 添加文件信息
    rpc addFile (fileInfo) returns (ma_reply) {}
    // 查找文件信息
    rpc searchFile (filePath) returns (searchReply) {}
    // 删除文件信息
    rpc delectFile (filePath) returns (ma_reply) {}
}

/*******************************/
/*         PARAMETER           */
/*******************************/
message serverInfo { // 数据服务器id,ip,port信息
    int32 id = 1;
    string ip = 2;
    int32 port = 3;
}

message serverId{
    int32 id = 1;
}

message fileInfo{
    string rootPath = 1;
    int32 serverId = 2;
    string filePath = 3;
}

message filePath{
    string filePath = 1;
}

message serverpath { // 数据服务器id信息
    string path = 1;
}

message empty { // 空信息
    int32 e = 1;
}

message lockInfo { // 包含客户端id和文件路径
    int32 clientId = 1;
    string filePath = 2;
}

/*******************************/
/*         RETURN_VALUE        */
/*******************************/
message ma_reply { // 操作是否成功
    bool done = 1;
}

message serverList { // 返回在线服务器信息列表
    repeated serverInfo list = 1;
}

message lockReply { // 返回文件上锁信息
    bool done = 1;
    string info = 2;
}

message searchReply{ // 返回服务器地址
    string path = 1;
    int32 serverId = 2;
    int32 done = 3;
}